### PR TITLE
CELDEV-994 - add 'defer' and 'async' support to JSScriptService calls

### PR DESCRIPTION
**Pull request by fpichler**
https://synjira.atlassian.net/browse/CELDEV-994

### CODE REVIEW

File: src/main/webapp/templates/celMacros/loadBackend.vm [hunk 1]
```diff
@@ -2,30 +2,30 @@
...
+##  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/adminUi/aspects.js'))
+##  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/adminUi/window.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/adminUi/admininterface.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/layoutSupport/Outliner.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':layouts/layoutEditor/outlineCells.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/contextMenu/contextmenu-hover.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/contextMenu/contextmenu.js'))
+##   $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/validation.js') ## always loaded in includeDefaultHeader.vm
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/adminUi/menubar.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/supportLink/supportLink.js'))
```
Comment (msladek) [inline src/main/webapp/templates/celMacros/loadBackend.vm:20]: is the builder intended to be reusable? does `includeExtJsFile` modify the builders' state?
if yes & no -> approve

File: src/main/webapp/templates/celMacros/DarkBox.vm [hunk 2]
```diff
@@ -12,12 +12,14 @@
...
+  #set($extJsParamDefer = $!jsService.createExtJSParam().setAction('file').setLoadMode($jsLoadModeDefer))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/yui/utilities/utilities.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/yui/dom/dom-min.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/yui/yahoo-dom-event/yahoo-dom-event.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/yui/animation/animation-min.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/yui/dragdrop/dragdrop-min.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/yui/element/element-min.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/yui/button/button-min.js'))
+  $!jsService.includeExtJsFile($extJsParamDefer.setJsFile(':celJS/yui/container/container-min.js'))
 #end
```
Comment (msladek) [inline src/main/webapp/templates/celMacros/DarkBox.vm]:
```suggestion
#end

```

File: src/main/webapp/resources/celJS/validation.js [hunk 3]
```diff
@@ -26,405 +26,408 @@
...
+      elem.select('form.celAddValidationToForm').each(Validation.addValidationToForm);
+    }
+  };
+  
+  document.addEventListener('DOMContentLoaded', function() {
+    $(document.body).observe('celValidation:addValidation',
+        Validation.addValidationToFormListener);
+    $(document.body).fire('celValidation:addValidation');
+  });
+})(window);
```
Comment (msladek) [inline src/main/webapp/resources/celJS/validation.js]:
```suggestion
})(window);

```

File: src/main/webapp/resources/celJS/validation.js [hunk 4]
```diff
@@ -26,405 +26,408 @@
...
+          const options = p.getElementsByTagName('INPUT');
+          return $A(options).any(function(elm) {
+            return $F(elm);
+          });
+        }],
+    ['validate-multiple-required', null, function (v, elm) {
+        const nr = elm.className.replace(/^.*multipleIs(\d+)( .*|)$/g, '$1');
+        const p = elm.up('.validation-one-required-field-wrapper') || elm.parentNode;
+        const options = p.getElementsByTagName('INPUT');
+        return nr == $A(options).findAll(function(elm) {
```
Comment (msladek) [inline src/main/webapp/resources/celJS/validation.js:372]: replace all `findAll` with `filter`

File: src/main/webapp/resources/celJS/validation.js [hunk 5]
```diff
@@ -26,405 +26,408 @@
...
+          if(typeof Effect == 'undefined') {
+            advice.style.display = 'block';
+          } else {
+            new Effect.Appear(advice, {duration : 1 });
+          }
+        }
+        elm[prop] = true;
+        elm.removeClassName('validation-passed');
+        elm.addClassName('validation-failed');
+        const elmWrapper = elm.up('.validation-field-wrapper');
```
Comment (msladek) [inline src/main/webapp/resources/celJS/validation.js:211]: replace all `up` with `clostest`

File: src/main/webapp/resources/celJS/validation.js [hunk 6]
```diff
@@ -26,405 +26,408 @@
...
+  
+  Object.extend(window.Validation, {
+    validate : function(elm, options){
+      options = Object.extend({
+        useTitle : false,
+        onElementValidate : function(result, elm) {}
+      }, options || {});
+      elm = $(elm);
+      const cn = elm.classNames();
+      return result = cn.all(function(value) {
```
Comment (msladek) [inline src/main/webapp/resources/celJS/validation.js:171]: replace all `all` with `every`

File: src/main/webapp/resources/celJS/validation.js [hunk 7]
```diff
@@ -26,405 +26,408 @@
...
-          advice.style.display = 'block';
-        } else {
-          new Effect.Appear(advice, {duration : 1 });
+        return false;
+      } else {
+        const advice = Validation.getAdvice(name, elm);
+        if(advice != null) advice.hide();
+        elm[prop] = '';
+        elm.removeClassName('validation-failed');
+        elm.addClassName('validation-passed');
```
Comment (msladek) [inline src/main/webapp/resources/celJS/validation.js:222]: replace all `remove/addClassName` with `classList` manipulations

File: src/main/webapp/resources/celJS/validation.js [hunk 8]
```diff
@@ -26,405 +26,408 @@
...
+          // [$].##
+          return Validation.get('IsEmpty').test(v) ||  /^\$?\-?([1-9]{1}[0-9]{0,2}(\,[0-9]{3})*(\.[0-9]{0,2})?|[1-9]{1}\d*(\.[0-9]{0,2})?|0(\.[0-9]{0,2})?|(\.[0-9]{1,2})?)$/.test(v);
+        }],
+    ['validate-selection', null, function(v, elm){
+          return elm.options ? elm.selectedIndex > 0 : !Validation.get('IsEmpty').test(v);
+        }],
+    ['validate-one-required', null, function (v, elm) {
+          const p = elm.up('.validation-one-required-field-wrapper') || elm.parentNode;
+          const options = p.getElementsByTagName('INPUT');
+          return $A(options).any(function(elm) {
```
Comment (msladek) [inline src/main/webapp/resources/celJS/validation.js:364]: `any` -> `some`

**Review by msladek — CHANGES_REQUESTED** [review]

File: src/main/webapp/templates/celMacros/loadBackend.vm (hunk 1 above)
Comment (fpichler) [reply src/main/webapp/templates/celMacros/loadBackend.vm:20]: builder is reusable. Added javadoc to the builder class https://github.com/celements/celements-core/pull/296/files

**Comment by fpichler** [comment]
Added new issue for refactoring/rewriting validation.js
https://synjira.atlassian.net/browse/CELDEV-1003

**Review by msladek — APPROVED** [review]

**Pull request merged**